timemachine/fe/single_topology_v3.py

```diff
@@ -735,3 +743,191 @@ def setup_intermediate_state(self, lamb):
+            i) at lambda = 0, the dummy atoms of mol_a are fully interacting, and become non-interacting at lambda = 1
+            ii) at lambda = 0, the dummy atoms of mol_b are non-interacting, and become fully interacting at lambda = 1
+            iii) the core atoms have parameters interpolated from mol_a's qlj to mol_b's qlj.
+        3) guest-guest interactions use pre-computed nonbonded interactions implemented as a pairlist.
+
+        Parameters
+        ----------
+        host_system: VacuumSystem
+            Parameterized system of the host
+
+        lamb: float
+            Which lambda value we want to generate the combined system.
+
+        Returns
+        -------
+        7-tuple of potentials
+            bond, angle, torsion, chiral_atom, chiral_bond, guest_nonbonded_precomputed, full_nonbonded
+
+        """
+
+        guest_system = self.setup_intermediate_state(lamb=lamb)
+
+        host_bond = host_system.bond
+        host_angle = host_system.angle
+        host_torsion = host_system.torsion
+        host_nonbonded = host_system.nonbonded
+
+        num_host_atoms = host_nonbonded.params.shape[0]
+
+        guest_bond = guest_system.bond
+        guest_angle = guest_system.angle
+        guest_torsion = guest_system.torsion
+        guest_nonbonded = guest_system.nonbonded
+        guest_chiral_atom = guest_system.chiral_atom
+        guest_chiral_bond = guest_system.chiral_bond
+
+        guest_chiral_atom_idxs = np.array(guest_chiral_atom.get_idxs(), dtype=np.int32) + num_host_atoms
+        guest_chiral_atom.set_idxs(guest_chiral_atom_idxs)
+
+        guest_chiral_bond_idxs = np.array(guest_chiral_bond.get_idxs(), dtype=np.int32) + num_host_atoms
+        guest_chiral_bond.set_idxs(guest_chiral_bond_idxs)
+
+        guest_nonbonded_idxs = np.array(guest_nonbonded.get_idxs(), dtype=np.int32) + num_host_atoms
+        guest_nonbonded.set_idxs(guest_nonbonded_idxs)
+
+        combined_bond_idxs = np.concatenate([host_bond.get_idxs(), guest_bond.get_idxs() + num_host_atoms])
+        combined_bond_params = np.concatenate([host_bond.params, guest_bond.params])
+        combined_bond = potentials.HarmonicBond(combined_bond_idxs).bind(combined_bond_params)
+
+        combined_angle_idxs = np.concatenate([host_angle.get_idxs(), guest_angle.get_idxs() + num_host_atoms])
+        combined_angle_params = np.concatenate([host_angle.params, guest_angle.params])
+        combined_angle = potentials.HarmonicAngle(combined_angle_idxs).bind(combined_angle_params)
+
+        if host_torsion:
+            combined_torsion_idxs = np.concatenate([host_torsion.get_idxs(), guest_torsion.get_idxs() + num_host_atoms])
+            combined_torsion_params = np.concatenate([host_torsion.params, guest_torsion.params])
+        else:
+            combined_torsion_idxs = np.array(guest_torsion.get_idxs(), dtype=np.int32) + num_host_atoms
+            combined_torsion_params = np.array(guest_torsion.params, dtype=np.float64)
```

Comment:
It's not obvious to me why a special case is needed for torsions. Worth adding a comment?

Reply:
solvent host doesn't have torsions, complex host does. will add comment